servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/DelayedSubscription.java

```diff
@@ -18,28 +18,26 @@
 import io.servicetalk.concurrent.PublisherSource.Subscription;
 
 import java.util.concurrent.atomic.AtomicLongFieldUpdater;
-import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 import javax.annotation.Nullable;
 
-import static io.servicetalk.concurrent.internal.ConcurrentSubscription.wrap;
-import static io.servicetalk.concurrent.internal.EmptySubscription.EMPTY_SUBSCRIPTION;
-import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
+import static io.servicetalk.concurrent.internal.FlowControlUtils.addWithOverflowProtection;
 import static java.lang.Long.MIN_VALUE;
-import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
 
 /**
  * A {@link Subscription} which serves as a placeholder until the "real" {@link Subscription} is available.
  */
 public final class DelayedSubscription implements Subscription {
-    private static final AtomicReferenceFieldUpdater<DelayedSubscription, Subscription> currentUpdater =
-            newUpdater(DelayedSubscription.class, Subscription.class, "current");
+    private static final long SUBSCRIPTION_SETTING = MIN_VALUE;
+    private static final long SUBSCRIPTION_SET = MIN_VALUE + 1;
+    private static final long SUBSCRIPTION_CANCEL_PENDING = MIN_VALUE + 2;
+    private static final long INVALID_GREATEST_VALUE = SUBSCRIPTION_CANCEL_PENDING;
     private static final AtomicLongFieldUpdater<DelayedSubscription> requestedUpdater =
-            AtomicLongFieldUpdater.newUpdater(DelayedSubscription.class, "requested");
+            newUpdater(DelayedSubscription.class, "requested");
 
-    @SuppressWarnings("unused")
     @Nullable
-    private volatile Subscription current;
-    @SuppressWarnings({"unused", "FieldCanBeLocal"})
+    private Subscription subscription;
     private volatile long requested;
```

Comment:
If you merge `requested` and `Subscription` fields in a single `Object state` then you can simplify the state-machine and reduce fields here.

(`requested` handling may be simplified, I haven't thought about it, but this gives the idea of what I am suggesting)

```java
public final class DelayedSubscription implements Subscription {
    private static final AtomicReferenceFieldUpdater<DelayedSubscription, Object> stateUpdater =
            AtomicReferenceFieldUpdater.newUpdater(DelayedSubscription.class, Object.class, "state");
    private static final Object STATE_CANCELLED = new Object();
    private static final Object STATE_TERMINATED = new Object();
    private volatile Object state = 0L;

    /**
     * Set the delayed {@link Subscription}. This method can only be called a single time and
     * subsequent calls will result in {@link #cancel()} being called on {@code delayedSubscription}.
     *
     * @param delayedSubscription The delayed {@link Subscription}.
     */
    public void delayedSubscription(Subscription delayedSubscription) {
        requireNonNull(delayedSubscription);
        for (;;) {
            final Object cState = state;
            if (cState == STATE_TERMINATED || cState instanceof Subscription) {
                delayedSubscription.cancel();
                return;
            }
            if (cState == STATE_CANCELLED && stateUpdater.compareAndSet(this, cState, STATE_TERMINATED)) {
                delayedSubscription.cancel();
                return;
            }

            if (cState instanceof Long && stateUpdater.compareAndSet(this, cState, 0L)) {
                long requested = (Long) cState;
                if (requested > 0 || requested == MIN_VALUE) {
                    delayedSubscription.request(requested);
                }
                if (stateUpdater.compareAndSet(this, 0L, delayedSubscription)) {
                    break;
                }
            }
        }
    }

    @Override
    public void request(long n) {
        for (;;) {
            final Object cState = state;
            if (cState == STATE_CANCELLED || cState == STATE_TERMINATED) {
                return;
            }
            if (cState instanceof Subscription) {
                Subscription s = (Subscription) cState;
                s.request(n);
                return;
            }
            if (cState instanceof Long) {
                Long prevReq = (Long) cState;
                if (n > 0 &&
                        stateUpdater.compareAndSet(this, cState, addWithOverflowProtectionIfNotNegative(prevReq, n))
                        || n <= 0 && stateUpdater.compareAndSet(this, cState, MIN_VALUE)) {
                    break;
                }
            }
        }
    }

    @Override
    public void cancel() {
        for (;;) {
            final Object cState = state;
            if (cState == STATE_CANCELLED || cState == STATE_TERMINATED) {
                return;
            }
            if (cState instanceof Subscription) {
                Subscription s = (Subscription) cState;
                s.cancel();
                return;
            }
            if (cState instanceof Long && stateUpdater.compareAndSet(this, cState, STATE_CANCELLED)) {
                return;
            }
        }
    }
}
```

Reply:
lets consider this alternative approach as a followup PR. the impacts of boxing/unboxing and additional object allocation need to be considered/evaluated. the motivation for this PR was avoiding intermediate allocations.